striatum/bandit/linucb.py

```diff
@@ -1,99 +1,157 @@
+
+    @property
+    def linucb(self):
+        """The generator implementing the disjoint LINUCB algorithm.
+        """
         while True:
-            invA = np.linalg.pinv(A)
-            theta = np.dot(invA, b)
-
-            p = np.dot(theta, x) + \
-                self.alpha * np.sqrt(
-                    np.einsum('ij,ji->i', np.dot(x.T, invA), x))
-
-            # next context x and last reward
-            at = np.random.choice(np.where(p == np.max(p))[0])
-            x_new, reward = yield at
-
-            A = A + np.dot(x[:, at], x[:, at].T)
-            b = b + reward * x[:, at]
-
-            x = x_new
+            context = yield
+            context = np.matrix(context)
+            matrix_ainv_tmp = np.array(
+                [self._modelstorage.get_model()['matrix_ainv'][action] for action in self._actions])
+            theta_tmp = np.array([self._modelstorage.get_model()['theta'][action] for action in self._actions])
+
+            # The recommended action should maximize the Linear UCB.
+            score = {}
+            for action_idx in range(len(self._actions)):
+                score[action_idx] = np.dot(context[action_idx], theta_tmp[action_idx]) + self.alpha * np.sqrt(
+                    np.dot(np.dot(context[action_idx], matrix_ainv_tmp[action_idx]), context[action_idx].T))
+            action_max = self._actions[np.argmax(score.values())]
+            yield action_max
+
+        raise StopIteration
 
     def get_action(self, context):
         """Return the action to perform
 
-        Parameters
-        ----------
-        context : {array-like, None}
-            The context of current state, None if no context avaliable.
-
-        Returns
-        -------
-        history_id : int
-            The history id of the action.
-
-        action : Actions object
-            The action to perform.
+            Parameters
+            ----------
+            context : {matrix-like, None}
+            The context of all actions at the current state. Row: action, Column: Context
+
+            Returns
+            -------
+            history_id : int
+                The history id of the actiself._actions_new = actionson.
+            action : Actions object
+                The action to perform.
         """
         if self.linucb_ is None:
-            self.linucb_ = self.linucb(context)
-        if self.last_reward is None:
-            raise ValueError("The last reward have not been passed in.")
-        action = self.linucb_.send(context, self.last_reward)
-
-        self.last_reward = None
-
-        history_id = self.storage.add_history(None, action, reward=None)
-        self.last_history_id = history_id
-
-        return history_id, action
+            self.linucb_ = self.linucb
+            self.linucb_.next()
+            action_max = self.linucb_.send(context)
+        else:
+            self.linucb_.next()
```

Comment:
It's python2 style generator, which will raise error in python 3+. Try ```six``` instead:
```python
import six

...
six.next(self.linucb_)
```

Document: https://pythonhosted.org/six/#six.next
StackOverflow: http://stackoverflow.com/questions/1073396/is-generator-next-visible-in-python-3-0